test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java

```diff
@@ -188,11 +197,79 @@ public void testSync() throws Exception {
+
         assertThat(log.getChannelsToSync().size()).isEqualTo(3);
 
         log.sync();
 
         assertThat(log.getChannelsToSync().size()).isEqualTo(0);
     }
+
+    @Test
+    public void testTrimUnknownAddress() throws Exception {
+        String logDir = getDirPath();
+        StreamLogFiles log = new StreamLogFiles(getDirPath(), false);
+        LogAddress logAddress = new LogAddress(0L, null);
+        log.trim(logAddress);
+        // Verify that the unknown address trim is not persisted
+        StreamLogFiles.SegmentHandle sh = log.getSegmentHandleForAddress(new LogAddress(0L, null));
+        assertThat(sh.getPendingTrims().size()).isEqualTo(0);
+    }
+
+    private void writeToLog(StreamLog log, Long addr) {
+        ByteBuf b = ByteBufAllocator.DEFAULT.buffer();
+        byte[] streamEntry = "Payload".getBytes();
+        Serializers.CORFU.serialize(streamEntry, b);
+        LogAddress address = new LogAddress(addr, null);
+        log.append(address, new LogData(DataType.DATA, b));
+    }
+
+    @Test
+    public void testTrim() throws Exception {
+        StreamLogFiles log = new StreamLogFiles(getDirPath(), false);
+
+        final int logChunk = StreamLogFiles.RECORDS_PER_LOG_FILE / 2;
+
+        // Write to the addresses then trim the addresses that span two log files
+        for (long x = 0; x < logChunk; x++) {
+            writeToLog(log, x);
+        }
+
+        // Verify that an incomplete log segment isn't compacted
+        for (long x = 0; x < logChunk; x++) {
+            LogAddress logAddress = new LogAddress(x, null);
+            log.trim(logAddress);
+        }
+
+        log.compact();
+
+        StreamLogFiles.SegmentHandle sh = log.getSegmentHandleForAddress(new LogAddress((long) logChunk, null));
+
+        assertThat(logChunk).isGreaterThan(StreamLogFiles.TRIM_THRESHOLD);
+        assertThat(sh.getPendingTrims().size()).isEqualTo(logChunk);
+        assertThat(sh.getTrimmedAddresses().size()).isEqualTo(0);
+
+        // Fill the rest of the log segment and compact
+        for (long x = logChunk; x < logChunk * 2; x++) {
+            writeToLog(log, x);
+        }
+
+        // Verify the pending trims are compacted after the log segment has filled
+        File file = new File(sh.getFileName());
+        long sizeBeforeCompact = file.length();
+        log.compact();
+        file = new File(sh.getFileName());
+        long sizeAfterCompact = file.length();
+
+        assertThat(sizeAfterCompact).isLessThan(sizeBeforeCompact);
+
+        // Reload the segment handler and check that the first half of the segment has been trimmed
+        sh = log.getSegmentHandleForAddress(new LogAddress((long) logChunk, null));
+        assertThat(sh.getTrimmedAddresses().size()).isEqualTo(logChunk);
+        assertThat(sh.getKnownAddresses().size()).isEqualTo(logChunk);
+
+        for (long x = logChunk; x < logChunk * 2; x++) {
+            assertThat(sh.getKnownAddresses().contains(x)).isTrue();
+        }
+    }
```

Comment:
I guess I'd like to have several more tests:

1. trim an address which was not written yet (should have a consistent effect, I am not sure what it is :)

2. trim an un-written address and then write it (same)

3. stream-read of a trimmed address (should return trimmed exception)

4. new stream-write of a previous trimmed address (same) 

5. think of more ...